### PR TITLE
Verificacao numerica do gradiente

### PR DESCRIPTION
**Pull request by diegodimer**
- [x] erro médio (sem ser quadrático)
- [x] verificação numérica do gradiente
- [x] refactor pra usar np.matrix
- [x] bias junto na theta_list, porém ignorando em alguns momentos quando necessário

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -157,6 +164,50 @@
...
+		avg_gradients_numeric.append(np.array(gradients_list))
+		for i in linhas:
+			print(f'\t\t\t{i}')
+
+v1 = abs( (avg_b-avg_bias_numeric).sum())
+v2 = abs( (avg_g-avg_gradients_numeric).sum())
+print("--------------------------------------------")
+print("Verificando corretude dos gradientes com base nos gradientes numericos:")
+print(f"\t\tErro entre gradiente via backprop e gradiente numerico para Theta1: {'%.10f' % (v1[0].sum() + v2[0].sum())}")
+print(f"\t\tErro entre gradiente via backprop e gradiente numerico para Theta2: {'%.10f' % (v1[1].sum() + v2[1].sum())}")
```
Comment (diegodimer) [inline main.py]: deveria ser:
```
--------------------------------------------
Verificando corretude dos gradientes com base nos gradientes numericos:
	Erro entre gradiente via backprop e gradiente numerico para Theta1: 0.0000000011
	Erro entre gradiente via backprop e gradiente numerico para Theta2: 0.0000000525
```

mas ta printando:
```
--------------------------------------------
Verificando corretude dos gradientes com base nos gradientes numericos:
                Erro entre gradiente via backprop e gradiente numerico para Theta1: 0.0000000002
                Erro entre gradiente via backprop e gradiente numerico para Theta2: 0.0000000001
```

deve ter alguma formula pra isso que a gente não tá ciente?
@ecpaim

File: main.py [hunk 2]
```diff
@@ -13,57 +13,72 @@
...
 
+def _normalize_df(df):
+	for column in df:
+		max_col = df[column].max()
+		min_col = df[column].min()
+		if max_col == min_col: #divisão por 0 faz dar NaN
+			df[column] = 1.0
+		else:
+			df[column] = (df[column] - min_col) / (max_col - min_col)
+		print(column)
```
Comment (diegodimer) [inline main.py:31]: se normaliza o conjunto de benchmark ele nao bate nenhum numero

File: main.py (hunk 1 above)
Comment (ecpaim) [reply main.py]: Eu tenho quase certeza que nesse caso não tem problema, porque essa parte é pra minimizar o erro.

Então se o algoritmo que tu escreveu tá dando um erro menor que o exemplo dele, quer dizer que ele consegue aproximar o gradiente melhor ainda (afinal isso daqui é só uma heurística). O que tá mais pra uma vantagem do que um erro.

Resumindo, acho que não precisa mudar nada.

File: neuralNetwork.py [hunk 3]
```diff
@@ -148,59 +146,41 @@ def backpropagation(self, test_set, debug = False):
...
-
-
-
-
-
-
-
-
+                for i in range(len(self.theta_list)):                  
+                    self.theta_list[i] = self.theta_list[i] - self.learning_rate * (gradients[i] + self.reg_factor*self.theta_list[i])
```
Comment (diegodimer) [inline neuralNetwork.py:183]: sera que não tá faltando um J aqui?

Comment (diegodimer) [inline neuralNetwork.py:183]: image
